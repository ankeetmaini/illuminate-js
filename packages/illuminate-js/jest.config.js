module.exports = {
    rootDir: __dirname,
    testMatch: ['**/?(*.)(spec|test).ts?(x)'],
    moduleFileExtensions: ['ts', 'tsx', 'js'],
    roots: ['<rootDir>/src'],
    transform: {
        '^.+\\.tsx?$': 'ts-jest'
    },
    collectCoverageFrom: ['<rootDir>/src/**/*.{ts,tsx}']
};
