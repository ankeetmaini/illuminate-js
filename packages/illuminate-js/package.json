{
  "name": "illuminate-js",
  "version": "1.0.0-alpha.1",
  "description": "Syntax Highlighting for NodeJS",
  "main": "./lib/index.js",
  "module": "./esm/index.js",
  "typings": "./lib/index.d.ts",
  "files": [
    "esm",
    "lib",
    "themes"
  ],
  "scripts": {
    "test": "jest --no-cache",
    "coverage": "jest --no-cache --coverage",
    "build": "yarn run build:idx && yarn run build:cjs && yarn run build:esm",
    "build:idx": "node scripts/build-index.js",
    "build:esm": "rimraf ./esm && tsc --target es6 --module es6 --outDir ./esm",
    "build:cjs": "rimraf ./lib && tsc --target es6 --module commonjs --outDir ./lib",
    "prepublishOnly": "yarn test && yarn run build"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/vkbansal/illuminate-js.git"
  },
  "author": "Vivek Kumar Bansal <contact@vkbansal.me>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/vkbansal/illuminate-js/issues"
  },
  "homepage": "https://github.com/vkbansal/illuminate-js#readme",
  "devDependencies": {
    "@types/jest": "^21.1.8",
    "jest": "^22.0.4",
    "rimraf": "^2.6.2",
    "ts-jest": "^22.0.0",
    "typescript": "^2.6.1"
  },
  "jest": {
    "testMatch": [
      "**/?(*.)(spec|test).ts?(x)"
    ],
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js"
    ],
    "roots": [
      "<rootDir>/src"
    ],
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "collectCoverageFrom": [
      "**/*.{ts,tsx}",
      "!**/__tests__/**"
    ],
    "mapCoverage": true
  }
}
