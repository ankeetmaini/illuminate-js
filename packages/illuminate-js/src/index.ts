import {
    TokenObject,
    TokenTypes,
    tokenize,
    Definition,
    Plugin,
    addPlugin,
    highlight,
    addLanguage,
    getLanguage
} from './illuminate';

export {
    TokenObject,
    TokenTypes,
    tokenize,
    Definition,
    Plugin,
    addPlugin,
    highlight,
    addLanguage,
    getLanguage
};
