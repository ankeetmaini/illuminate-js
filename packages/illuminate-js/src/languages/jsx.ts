import { javascript } from './javascript';
import { markup } from './markup';
import { clone, setIn, getIn, insertBefore } from '../utils';
import { Definition } from '../illuminate';

let jsx: Definition = clone(markup);

javascript.forEach((value, key) => jsx.set(key, value));

setIn(
    jsx,
    ['tag', 'pattern'],
    /<\/?[\w.:-]+\s*(?:\s+(?:[\w.:-]+(?:=(?:("|')(?:\\[\s\S]|(?!\1)[^\\])*\1|[^\s'">=]+|(?:\{\{?[^}]*\}?\})))?|\{\.{3}[a-z_$][\w$]*(?:\.[a-z_$][\w$]*)*\}))*\s*\/?>/i
);
setIn(
    jsx,
    ['tag', 'inside', 'attr-value', 'pattern'],
    /=(?!\{)(?:("|')(?:\\[\s\S]|(?!\1)[^\\])*\1|[^\s'">]+)/i
);

insertBefore(
    getIn<Definition>(jsx, ['tag', 'inside']),
    'attr-name',
    new Map([
        [
            'spread',
            {
                pattern: /\{\.{3}[a-z_$][\w$]*(?:\.[a-z_$][\w$]*)*\}/,
                inside: new Map([['punctuation', /\.{3}|[{}.]/], ['attr-value', /\w+/]])
            }
        ]
    ])
);

let jsxExpression = clone(jsx);
jsxExpression.delete('punctuation');

jsxExpression = insertBefore(
    jsxExpression,
    'operator',
    new Map([['punctuation', /=(?={)|[{}[\];(),.:]/]])
);

insertBefore(
    getIn(jsx, ['tag', 'inside']),
    'attr-value',
    new Map([
        [
            'script',
            {
                // Allow for one level of nesting
                pattern: /=(\{(?:\{[^}]*\}|[^}])+\})/i,
                inside: jsxExpression,
                alias: 'language-javascript'
            }
        ]
    ])
);

export { jsx };
