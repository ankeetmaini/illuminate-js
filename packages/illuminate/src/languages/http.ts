import { javascript } from './javascript';
import { markup } from './markup';
import { Definition, TokenTypes, TokenObject } from '../illuminate';
import { insertBefore } from '../utils';

const http: Definition = new Map<string, TokenTypes>([
    [
        'request-line',
        {
            pattern: /^(?:POST|GET|PUT|DELETE|OPTIONS|PATCH|TRACE|CONNECT)\shttps?:\/\/\S+\sHTTP\/[0-9.]+/m,
            inside: new Map([
                // HTTP Verb
                ['property', /^(?:POST|GET|PUT|DELETE|OPTIONS|PATCH|TRACE|CONNECT)\b/],
                // Path or query argument
                ['attr-name', /:\w+/]
            ])
        }
    ],
    [
        'response-status',
        {
            pattern: /^HTTP\/1.[01] \d+.*/m,
            inside: new Map([
                // Status, e.g. 200 OK
                [
                    'property',
                    {
                        pattern: /(^HTTP\/1.[01] )\d+.*/i,
                        lookbehind: true
                    }
                ]
            ])
        }
    ],
    // HTTP header name
    [
        'header-name',
        {
            pattern: /^[\w-]+:(?=.)/m,
            alias: 'keyword'
        }
    ]
]);

// Create a mapping of Content-Type headers to language definitions
const httpLanguages: Record<string, Definition> = {
    'application/json': javascript,
    'application/xml': markup,
    'text/xml': markup,
    'text/html': markup
};

// Insert each content type parser that has its associated language
// currently loaded.

const options = new Map<string, TokenObject>();

for (const contentType in httpLanguages) {
    if (!httpLanguages.hasOwnProperty(contentType)) continue;

    options.set(contentType, {
        pattern: new RegExp(
            `(content-type:\\s*${contentType}[\\w\\W]*?)(?:\\r?\\n|\\r){2}[\\w\\W]*`,
            'i'
        ),
        lookbehind: true,
        inside: new Map<string, TokenTypes>([['rest', httpLanguages[contentType] as TokenTypes]])
    });
}

insertBefore(http, 'header-name', options);

export { http };
