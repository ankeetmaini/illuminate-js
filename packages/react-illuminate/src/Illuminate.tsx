import * as React from 'react';
import * as cx from 'classnames';
import { tokenize, Definition, TokenObject } from 'illuminate-js';

import { TokenElement, IlluminateToken, SingleOrArray, CustomClasses } from './Token';

export const languages = new Map<string, Definition>();

export { Definition, TokenObject };

export interface IlluminateProps {
    /**
     * The text to be highlighted
     */
    children: string;
    /**
     * The language definition, to be used to highlight the text.
     * This must match with one of the language names, used with `addLanguage`
     */
    lang: string;
    /**
     * Flag to toggle the addition of line-numbers to output.
     * **Note**: You might need to add appropriate styles for this to work.
     * See Illuminate Plugins for more details
     *
     * @default false
     */
    lineNumbers?: boolean;
    /**
     * Flag to toggle the addition of language name to output.
     * **Note**: You might need to add appropriate styles for this to work.
     * See Illuminate Plugins for more details
     *
     * @default false
     */
    showLanguage?: boolean;
    /**
     * By default, Illuminate uses fixed and generic class names.
     * Using this setting, you can cutomize those classes.
     * For example:
     *
     *   * You can namespace classes like `.illuminate--string` to avoid conflicts with existing CSS.
     *   * You can use [CSS Modules](https://github.com/css-modules/css-modules), like `.operator_93jsa`.
     *
     * @example
     *
     * {
     *      prefix: 'prefix-',
     *      map: {
     *          property: 'special-property',
     *          string: 'string_ch29s',
     *          operator: 'operator_93jsa'
     *      }
     * }
     */
    customClasses?: CustomClasses;
}

export class Illuminate extends React.PureComponent<IlluminateProps> {
    /**
     * Add a language definition to a react-illuminate.
     * You can add same definition multiple times with different names
     *
     * @param name  The name of the language being added.
     *              This name will be used by `highlight` function.
     * @param def   The language `Definition` to be added.
     *
     * @example
     * import { Illuminate } from 'react-illuminate';
     * import { javascript } from 'illuminate-js/lib/languages';
     *
     * Illuminate.addLanguage('javascript', javascript);
     * Illuminate.addLanguage('js', javascript);
     */
    static addLanguage(name: string, def: Definition) {
        languages.set(name, def);
    }

    /**
     * Returns the language definition which was added using `Illuminate.addLanguage`.
     * Returns the `Definition` if found, else returns `undefined`.
     *
     * @param name The name of the language definition
     *
     * @example
     * import { Illuminate } from 'react-illuminate';
     * import { javascript } from 'illuminate-js/lib/languages';
     *
     * Illuminate.addLanguage('javascript', javascript);
     * Illuminate.addLanguage('js', javascript);
     *
     * Illuminate.getLanguage('js'); // returns javascript definition
     */
    static getLanguage(name: string): Definition | undefined {
        return languages.get(name);
    }

    /**
     * Delete the language definition which was added using `Illuminate.addLanguage`.
     *
     * @param name The name of the language definition
     *
     *
     * @example
     * import { Illuminate } from 'react-illuminate';
     * import { javascript } from 'illuminate-js/lib/languages';
     *
     * Illuminate.addLanguage('javascript', javascript);
     * Illuminate.addLanguage('js', javascript);
     *
     * Illuminate.getLanguage('js'); // returns javascript definition
     */
    static deleteLanguage(name: string): void {
        languages.delete(name);
    }

    renderLineNumbers(code: string) {
        const match = code.match(/\n(?!$)/g);
        const linesNum = match ? match.length + 1 : 1;
        let lines = [];

        for (let i = 0; i < linesNum; i++) {
            lines.push(<span key={`line-${i}`} className="line-number" />);
        }

        return <span className="line-numbers-rows">{lines}</span>;
    }

    render() {
        const { lang, children } = this.props;
        let token: string | SingleOrArray<string | IlluminateToken> = children;

        const grammar = Illuminate.getLanguage(lang || '');

        if (!grammar) {
            return children;
        }

        token = tokenize(children, grammar);

        const className = cx(`language-${lang}`, {
            'line-numbers': this.props.lineNumbers,
            'show-language': this.props.showLanguage
        });

        return (
            <pre className={className}>
                <code>
                    {this.props.showLanguage && <span className="show-language--lang">{lang}</span>}
                    <TokenElement customClasses={this.props.customClasses} token={token} />
                    {this.props.lineNumbers && this.renderLineNumbers(children)}
                </code>
            </pre>
        );
    }
}
