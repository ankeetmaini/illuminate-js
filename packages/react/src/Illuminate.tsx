import * as React from 'react';
import { tokenize, getLanguage } from 'illuminate-js';

import { TokenElement, Token, SingleOrArray } from './Token';

export interface IlluminateProps {
    children: string;
    lang?: string;
}

export default class Illuminate extends React.Component<IlluminateProps> {
    shouldComponentUpdate(nextProps: IlluminateProps) {
        return this.props.children !== nextProps.children || this.props.lang !== nextProps.lang;
    }

    render() {
        const { lang, children } = this.props;
        let token: string | SingleOrArray<string | Token> = children;

        const grammar = getLanguage(lang || '');

        if (!grammar) {
            return children;
        }

        token = tokenize(children, grammar);

        return (
            <pre className={`language-${lang}`}>
                <code>
                    <TokenElement token={token} />
                </code>
            </pre>
        );
    }
}
