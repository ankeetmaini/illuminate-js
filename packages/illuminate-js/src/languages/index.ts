// This is an autogenerated file, please do not change anything here

import { apacheconf } from './apacheconf';
import { bash } from './bash';
import { clike } from './clike';
import { css } from './css';
import { diff } from './diff';
import { haskell } from './haskell';
import { http } from './http';
import { ini } from './ini';
import { javascript } from './javascript';
import { json } from './json';
import { jsx } from './jsx';
import { less } from './less';
import { makefile } from './makefile';
import { markdown } from './markdown';
import { markup } from './markup';
import { matlab } from './matlab';
import { php } from './php';
import { python } from './python';
import { ruby } from './ruby';
import { scss } from './scss';
import { sql } from './sql';
import { typescript } from './typescript';
import { yaml } from './yaml';

export {
    apacheconf,
    bash,
    clike,
    css,
    diff,
    haskell,
    http,
    ini,
    javascript,
    json,
    jsx,
    less,
    makefile,
    markdown,
    markup,
    matlab,
    php,
    python,
    ruby,
    scss,
    sql,
    typescript,
    yaml,
    markup as html,
    javascript as js,
    json as json5,
    markdown as md,
    typescript as ts
};
