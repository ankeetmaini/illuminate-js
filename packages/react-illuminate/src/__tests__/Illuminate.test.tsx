import * as React from 'react';
import { addLanguage } from 'illuminate-js';
import { clike } from 'illuminate-js/lib/languages/clike';
import { shallow } from 'enzyme';

import { Illuminate } from '../Illuminate';

describe('<Illuminate /> test', () => {
    beforeAll(() => {
        addLanguage('clike', clike);
    });

    it('should highlight given text', () => {
        const component = shallow(<Illuminate lang="clike">if (a > b)</Illuminate>);
        expect(component).toMatchSnapshot();
    });
});
