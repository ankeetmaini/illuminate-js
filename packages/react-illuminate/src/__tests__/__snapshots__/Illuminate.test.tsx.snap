// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Illuminate /> test should highlight given text 1`] = `
<pre
  className="language-clike"
>
  <code>
    <TokenElement
      token={
        Array [
          Token {
            "alias": undefined,
            "content": "if",
            "type": "keyword",
          },
          " ",
          Token {
            "alias": undefined,
            "content": "(",
            "type": "punctuation",
          },
          "a ",
          Token {
            "alias": undefined,
            "content": ">",
            "type": "operator",
          },
          " b",
          Token {
            "alias": undefined,
            "content": ")",
            "type": "punctuation",
          },
        ]
      }
    />
  </code>
</pre>
`;
